### ISSUE TITLE
No diagnostic is reported for illegal macro definitions such as "noreturn" and "carries_dependency"

### ISSUE THREAD

**Issue by romanova-ekaterina** [llvm/llvm-project#92196]
Consider the testcase like that:
```cpp
#define override 1 
#define noreturn 2
#define final 3 
#define carries_dependency 4

int foo(int a)
{
  return (a == override + noreturn + final + carries_dependency);
}
```

When compiling it with clang compiler, I see that the diagnostic is issued for 2 out 4 macros (override and final), which is inconsistent.
 
I could see the argument why no error should be reported. I could see an argument why all four errors should be reported. But not reporting diagnostic only for 2 out of 4 macros that falls into the same category is suprising.

(note: the same behavior happens with c++14, c++17, c++20 standard)

```
clang -c -Wpedantic -Wall -std=c++11 test.cpp
test.cpp:2:9: warning: keyword is hidden by macro definition [-Wkeyword-macro]
#define override 1
        ^
test.cpp:4:9: warning: keyword is hidden by macro definition [-Wkeyword-macro]
#define final 3
        ^
2 warnings generated.
=======================
```

On one hand, https://eel.is/c++draft/macro.names#2 says:
"A translation unit shall not #define or #undef names lexically identical to keywords, to the identifiers listed in Table 4, or to the attribute-tokens described in [dcl.attr], except that the names likely and unlikely may be defined as function-like macros ([cpp.replace]).
"Table 4" is: "final import module override".
Redefining attribute-tokens is also not allowed.
https://eel.is/c++draft/dcl.attr; "carries_dependency" and "noreturn" are both in there (attribute tokens).

The use of the language "shall not" means that the program is ill-formed and requires a diagnostic, according to cppreference.

On the other hand, it might be considered as and undefined bahavior.
There is a 10 years old thread on llvm.org explaining why error should not be reported.
[However, it doesn't address my inconsistency question]
https://lists.llvm.org/pipermail/cfe-dev/2013-June/030283.html

Summary:
If Clang diagnoses a redefinition some attribute names such as "final" and "override", then failing to do so for "carries_dependency" and "noreturn" seems like a Clang bug.

**Comment by shafik** [comment]
It does not seem like other compilers complain about this: https://godbolt.org/z/EbrMjEG69

**Comment by zygoloid** [comment]
> The use of the language "shall not" means that the program is ill-formed and requires a diagnostic, according to cppreference.

The context is "[a] translation unit that includes a standard library header", and this is in library wording where "shall" doesn't have quite the same implication as in core wording. Note also that clang doesn't issue an error here on any of these cases even in `-pedantic-errors` mode.

I don't think it's inconsistent to warn on context-sensitive keywords but not on attribute names, given that attribute names aren't any kind of keyword, and attributes are still usable via `__decorated__` names for code that wants to be careful about such things.

**Comment by AaronBallman** [comment]
> attributes are still usable via __decorated__ names for code that wants to be careful about such things.

This is true for folks compiling with Clang, GCC, or EDG, but Microsoft currently does not provide a mechanism by which users can protect their attribute usages, [despite requests](https://github.com/llvm/llvm-project/pull/113765#issuecomment-2444874495).

> I don't think it's inconsistent to warn on context-sensitive keywords but not on attribute names, given that attribute names aren't any kind of keyword, and attributes are still usable via __decorated__ names for code that wants to be careful about such things.

From a language lawyer and compiler writer perspective, I agree. However, I think users can often think in terms like "these names are reserved" without thinking about it in terms of language grammar productions. So I'd be fine if we decided to issue a diagnostic for use of attribute names as macros so long as at least one standard library header is included. However, the include stack can be deep, so I worry about compile time overhead a little bit (perhaps there are enough "if a standard library header is included" rules in the standard that we could consider using  a bit somewhere to track "at least one standard library header has been included"?).

**Comment by AaronBallman** [comment]
I think we could call `Sema::getStdNamespace()` and if it returns nonnull, presume the namespace has been loaded, but we may have to look at whether it's been implicitly created or not (we have `Sema::getOrCreateStdNamespace()` which could have inserted the namespace without an include).

**Comment by a-tarasyuk** [comment]
The proposed [PR](https://github.com/llvm/llvm-project/pull/106036) extends `PPDirectives` to handle attribute names., at this stage, `getStdNamespace` isn’t available because it operates at the _Lex_ level. I looked into this further, and since `PPDirectives` lacks _AST_ context, one possible approach (though I don’t think it’s appropriate to include _sema_-related logic into `PPDirectives`) would be to analyze the token sequence to check if `namespace` `std` has been encountered, however, this approach seems unreliable. A viable alternative would be to move this check to the sema level, where the necessary context is available, though it would require additional effort...

**Comment by a-tarasyuk** [comment]
@AaronBallman @philnik777 I've updated the PR with draft changes that use existing hooks and rely on `getStdNamespace`. Could you please review them to see if this is a suitable direction to proceed with?

**Comment by a-tarasyuk** [comment]
@erichkeane https://github.com/llvm/llvm-project/pull/106036#issuecomment-2567868500 - the latest changes aim to address the https://github.com/llvm/llvm-project/issues/92196#issuecomment-2497896303, https://github.com/llvm/llvm-project/issues/92196#issuecomment-2497965263
